Unable to convert java.math.BigDecimal to Neo4j Value

## Guidelines

Please help me to resolve the issue if you can. 

I config-ed Neo4j Sink Connector to write the data from a Kafka topic to Neo4j. However I'm facing an issue that makes Sink Connector not working recently

```
Caused by: org.apache.kafka.connect.errors.DataException: Exception thrown while processing field 'target'
        at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:259)
        at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convert(AbstractConverter.java:78)
        at streams.kafka.connect.sink.Neo4jService.write(Neo4jService.kt:94)
        at streams.service.StreamsSinkService.writeWithCypherTemplate(StreamsSinkService.kt:36)
        at streams.service.StreamsSinkService.writeForTopic(StreamsSinkService.kt:42)
        at streams.kafka.connect.sink.Neo4jService$writeData$2$invokeSuspend$$inlined$flatMap$lambda$1.invokeSuspend(Neo4jService.kt:121)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
        at kotlinx.coroutines.DispatchedTask$DefaultImpls.run(Dispatched.kt:235)
        at kotlinx.coroutines.DispatchedContinuation.run(Dispatched.kt:81)
        at kotlinx.coroutines.scheduling.Task.run(Tasks.kt:94)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:732)
Caused by: org.neo4j.driver.v1.exceptions.ClientException: Unable to convert java.math.BigDecimal to Neo4j Value.
        at org.neo4j.driver.v1.Values.value(Values.java:130)
        at streams.kafka.connect.sink.ValueConverter.setValue(ValueConverter.kt:23)
        at streams.kafka.connect.sink.ValueConverter.setDecimalField(ValueConverter.kt:112)
        at streams.kafka.connect.sink.ValueConverter.setDecimalField(ValueConverter.kt:17)
        at com.github.jcustenborder.kafka.connect.utils.data.AbstractConverter.convertStruct(AbstractConverter.java:223)
        ... 12 more
```
I'm wondering how I can figure out the root cause and troubleshot the issue in this case since the log is only thing I have.

The avro schema generated automatically by Kafka Connect JDBC
```
{
  "type": "record",
  "name": "ConnectDefault",
  "namespace": "io.confluent.connect.avro",
  "fields": [
    {
      "name": "goal_id",
      "type": "string"
    },
    {
      "name": "publisher",
      "type": "string"
    },
    {
      "name": "lineitem_id",
      "type": "string"
    },
    {
      "name": "position_id",
      "type": {
        "type": "int",
        "connect.type": "int16"
      }
    },
    {
      "name": "indicator_name",
      "type": "string"
    },
    {
      "name": "target",
      "type": {
        "type": "bytes",
        "scale": 6,
        "precision": 64,
        "connect.version": 1,
        "connect.parameters": {
          "scale": "6"
        },
        "connect.name": "org.apache.kafka.connect.data.Decimal",
        "logicalType": "decimal"
      }
    },
    {
      "name": "result_val",
      "type": [
        "null",
        {
          "type": "bytes",
          "scale": 6,
          "precision": 64,
          "connect.version": 1,
          "connect.parameters": {
            "scale": "6"
          },
          "connect.name": "org.apache.kafka.connect.data.Decimal",
          "logicalType": "decimal"
        }
      ],
      "default": null
    },
    {
      "name": "tracking_value",
      "type": [
        "null",
        {
          "type": "bytes",
          "scale": 2,
          "precision": 64,
          "connect.version": 1,
          "connect.parameters": {
            "scale": "2"
          },
          "connect.name": "org.apache.kafka.connect.data.Decimal",
          "logicalType": "decimal"
        }
      ],
      "default": null
    },
    {
      "name": "tracking_pct",
      "type": [
        "null",
        {
          "type": "bytes",
          "scale": 2,
          "precision": 64,
          "connect.version": 1,
          "connect.parameters": {
            "scale": "2"
          },
          "connect.name": "org.apache.kafka.connect.data.Decimal",
          "logicalType": "decimal"
        }
      ],
      "default": null
    },
    {
      "name": "daily_target",
      "type": [
        "null",
        {
          "type": "bytes",
          "scale": 2,
          "precision": 64,
          "connect.version": 1,
          "connect.parameters": {
            "scale": "2"
          },
          "connect.name": "org.apache.kafka.connect.data.Decimal",
          "logicalType": "decimal"
        }
      ],
      "default": null
    },
    {
      "name": "is_disabled",
      "type": "boolean"
    },
    {
      "name": "metric_type",
      "type": [
        "null",
        "string"
      ],
      "default": null
    },
    {
      "name": "pg_field",
      "type": [
        "null",
        "string"
      ],
      "default": null
    },
    {
      "name": "refresh_sql_statement",
      "type": "string"
    },
    {
      "name": "last_updated",
      "type": {
        "type": "long",
        "connect.version": 1,
        "connect.name": "org.apache.kafka.connect.data.Timestamp",
        "logicalType": "timestamp-millis"
      }
    },
    {
      "name": "hash",
      "type": "string"
    }
  ]
}
```

## Expected Behavior (Mandatory)


## Actual Behavior (Mandatory)


## How to Reproduce the Problem

### Simple Dataset (where it's possibile)

```
//Insert here a set of Cypher statements that helps us to reproduce the problem

```


### Steps (Mandatory)

  1.
  1.
  1.

## Screenshots (where it's possibile)

## Specifications (Mandatory)

Currently used versions

### Versions

  - OS: AWS EC2
  - Neo4j: 3.5.14
  - Neo4j-Streams: 3.5.5

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -13,6 +13,20 @@ import java.util.concurrent.TimeUnit
 
 class Neo4jValueConverter: MapValueConverter<Value>() {
     private val log = LoggerFactory.getLogger(Neo4jValueConverter::class.java)
+    private val errorMsg = "Cannot convert %s field into Neo4j Type, it exceeds the Neo4j %s value, it will be converted into a String"
+
+    private fun parseBigDecimal(key: String?, value: BigDecimal?): Any {
+        val doubleValue = value!!.toDouble()
+        val fitsScale = doubleValue != Double.POSITIVE_INFINITY
+                && doubleValue != Double.NEGATIVE_INFINITY
+                && value.compareTo(BigDecimal.valueOf(doubleValue)) == 0
+        return if (fitsScale) {
+            doubleValue
+        } else {
+            log.warn(errorMsg.format(key, "Double"))
+            value.toPlainString()
+        }
+    }
 
     companion object {
         @JvmStatic private val UTC = ZoneId.of("UTC")
@@ -21,26 +35,13 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
     override fun convert(data: Any?): MutableMap<String, Value?> {
         val dataManipulatedForNeo4j = if(data is Map<*, *>)
             data.mapValues {
-                val value = it.value
-                val errorMsg = "Cannot convert ${it.key} key into Neo4j Type, it exceeds the Neo4j %s value, it will be converted into a String"
-                when (value) {
-                    is BigDecimal -> {
-                        val doubleValue = value.toDouble()
-                        val fitsScale = doubleValue != Double.POSITIVE_INFINITY
-                                && doubleValue != Double.NEGATIVE_INFINITY
-                                && value.compareTo(BigDecimal.valueOf(doubleValue)) == 0
-                        if (fitsScale) {
-                            doubleValue
-                        } else {
-                            log.warn(errorMsg.format("Double"))
-                            value.toPlainString()
-                        }
-                    }
+                when (val value = it.value) {
+                    is BigDecimal -> parseBigDecimal(it.key as String, value)
                     is BigInteger -> {
                         try {
                             value.longValueExact()
                         } catch (e: java.lang.ArithmeticException) {
-                            log.warn(errorMsg.format("Long"))
+                            log.warn(errorMsg.format(it.key, "Long"))
                             value.toString()
                         }
                     }
@@ -59,6 +60,10 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
 
     override fun newValue(): MutableMap<String, Value?> {
         return mutableMapOf()
+    }
+
+    override fun setDecimalField(result: MutableMap<String, Value?>?, fieldName: String?, value: BigDecimal?) {
+        setValue(result, fieldName, parseBigDecimal(fieldName, value))
     }
 
     override fun setTimestampField(result: MutableMap<String, Value?>?, fieldName: String?, value: Date?) {

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -2,17 +2,56 @@ package streams.kafka.connect.sink.converters
 
 import org.neo4j.driver.Value
 import org.neo4j.driver.Values
+import org.slf4j.LoggerFactory
+import java.math.BigDecimal
+import java.math.BigInteger
 import java.time.LocalTime
 import java.time.ZoneId
-import java.util.Date
+import java.util.*
 import java.util.concurrent.TimeUnit
 
 
 class Neo4jValueConverter: MapValueConverter<Value>() {
+    private val log = LoggerFactory.getLogger(Neo4jValueConverter::class.java)
 
     companion object {
         @JvmStatic private val UTC = ZoneId.of("UTC")
     }
+
+
+    override fun convert(data: Any?): MutableMap<String, Value?> {
+        val dataManipulatedForNeo4j = if(data is Map<*, *>)
+            data.mapValues {
+                val value = it.value
+                val errorMsg = "Cannot convert %s key into Neo4j Type, it exceeds the Neo4j %s value, it will be converted into a String"
+                when (value) {
+                    is BigDecimal -> {
+                        val doubleValue = value.toDouble()
+                        val fitsScale = doubleValue != Double.POSITIVE_INFINITY
+                                && doubleValue != Double.NEGATIVE_INFINITY
+                                && value.compareTo(BigDecimal.valueOf(doubleValue)) == 0
+                        if (fitsScale) {
+                            doubleValue
+                        } else {
+                            log.warn(errorMsg.format(it.key, "Double"))
+                            value.toPlainString()
+                        }
+                    }
+                    is BigInteger -> {
+                        try {
+                            value.longValueExact()
+                        } catch (e: java.lang.ArithmeticException) {
+                            log.warn(errorMsg.format(it.key, "Long"))
+                            value.toString()
+                        }
+                    }
+                    else -> value
+                }
+            } else data
+
+        return super.convert(dataManipulatedForNeo4j)
+    }
+
 
     override fun setValue(result: MutableMap<String, Value?>?, fieldName: String?, value: Any?) {
         if (result != null && fieldName != null) {

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -18,7 +18,6 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
         @JvmStatic private val UTC = ZoneId.of("UTC")
     }
 
-
     override fun convert(data: Any?): MutableMap<String, Value?> {
         val dataManipulatedForNeo4j = if(data is Map<*, *>)
             data.mapValues {
@@ -51,7 +50,6 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
 
         return super.convert(dataManipulatedForNeo4j)
     }
-
 
     override fun setValue(result: MutableMap<String, Value?>?, fieldName: String?, value: Any?) {
         if (result != null && fieldName != null) {

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -12,8 +12,6 @@ import java.util.concurrent.TimeUnit
 
 
 class Neo4jValueConverter : MapValueConverter<Value>() {
-    private val log = LoggerFactory.getLogger(Neo4jValueConverter::class.java)
-    private val errorMsg = "Cannot convert %s field into Neo4j Type, it exceeds the Neo4j %s value, it will be converted into a String"
 
     companion object {
         @JvmStatic
@@ -28,7 +26,6 @@ class Neo4jValueConverter : MapValueConverter<Value>() {
                     try {
                         value.longValueExact()
                     } catch (e: java.lang.ArithmeticException) {
-                        log.warn(errorMsg.format(it.key, "Long"))
                         value.toString()
                     }
                 } else {
@@ -54,15 +51,14 @@ class Neo4jValueConverter : MapValueConverter<Value>() {
     }
 
     override fun setDecimalField(result: MutableMap<String, Value?>?, fieldName: String?, value: BigDecimal?) {
-        val doubleValue = value!!.toDouble()
+        val doubleValue = value?.toDouble()
         val fitsScale = doubleValue != Double.POSITIVE_INFINITY
                 && doubleValue != Double.NEGATIVE_INFINITY
-                && value.compareTo(BigDecimal.valueOf(doubleValue)) == 0
+                && value?.compareTo(doubleValue?.let { BigDecimal.valueOf(it) }) == 0
         if (fitsScale) {
             setValue(result, fieldName, doubleValue)
         } else {
-            log.warn(errorMsg.format(fieldName, "Double"))
-            setValue(result, fieldName, value.toPlainString())
+            setValue(result, fieldName, value?.toPlainString())
         }
     }
 

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -11,7 +11,7 @@ import java.util.*
 import java.util.concurrent.TimeUnit
 
 
-class Neo4jValueConverter : MapValueConverter<Value>() {
+class Neo4jValueConverter: MapValueConverter<Value>() {
 
     companion object {
         @JvmStatic

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -2,20 +2,17 @@ package streams.kafka.connect.sink.converters
 
 import org.neo4j.driver.Value
 import org.neo4j.driver.Values
-import org.slf4j.LoggerFactory
 import java.math.BigDecimal
-import java.math.BigInteger
 import java.time.LocalTime
 import java.time.ZoneId
-import java.util.*
+import java.util.Date
 import java.util.concurrent.TimeUnit
 
 
 class Neo4jValueConverter: MapValueConverter<Value>() {
 
     companion object {
-        @JvmStatic
-        private val UTC = ZoneId.of("UTC")
+        @JvmStatic private val UTC = ZoneId.of("UTC")
     }
 
     override fun setValue(result: MutableMap<String, Value?>?, fieldName: String?, value: Any?) {

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -23,7 +23,7 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
         val dataManipulatedForNeo4j = if(data is Map<*, *>)
             data.mapValues {
                 val value = it.value
-                val errorMsg = "Cannot convert %s key into Neo4j Type, it exceeds the Neo4j %s value, it will be converted into a String"
+                val errorMsg = "Cannot convert ${it.key} key into Neo4j Type, it exceeds the Neo4j %s value, it will be converted into a String"
                 when (value) {
                     is BigDecimal -> {
                         val doubleValue = value.toDouble()
@@ -33,7 +33,7 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
                         if (fitsScale) {
                             doubleValue
                         } else {
-                            log.warn(errorMsg.format(it.key, "Double"))
+                            log.warn(errorMsg.format("Double"))
                             value.toPlainString()
                         }
                     }
@@ -41,7 +41,7 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
                         try {
                             value.longValueExact()
                         } catch (e: java.lang.ArithmeticException) {
-                            log.warn(errorMsg.format(it.key, "Long"))
+                            log.warn(errorMsg.format("Long"))
                             value.toString()
                         }
                     }

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -18,28 +18,6 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
         private val UTC = ZoneId.of("UTC")
     }
 
-    override fun convert(data: Any?): MutableMap<String, Value?> {
-        val dataManipulatedForNeo4j = if (data is Map<*, *>) {
-            data.mapValues {
-                val value = it.value
-                if (value is BigInteger) {
-                    try {
-                        value.longValueExact()
-                    } catch (e: java.lang.ArithmeticException) {
-                        value.toString()
-                    }
-                } else {
-                    value
-                }
-            }
-        } else {
-            data
-        }
-
-        return super.convert(dataManipulatedForNeo4j)
-    }
-
-
     override fun setValue(result: MutableMap<String, Value?>?, fieldName: String?, value: Any?) {
         if (result != null && fieldName != null) {
             result[fieldName] = Values.value(value) ?: Values.NULL

--- a/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
+++ b/kafka-connect-neo4j/src/main/kotlin/streams/kafka/connect/sink/converters/Neo4jValueConverter.kt
@@ -11,46 +11,37 @@ import java.util.*
 import java.util.concurrent.TimeUnit
 
 
-class Neo4jValueConverter: MapValueConverter<Value>() {
+class Neo4jValueConverter : MapValueConverter<Value>() {
     private val log = LoggerFactory.getLogger(Neo4jValueConverter::class.java)
     private val errorMsg = "Cannot convert %s field into Neo4j Type, it exceeds the Neo4j %s value, it will be converted into a String"
 
-    private fun parseBigDecimal(key: String?, value: BigDecimal?): Any {
-        val doubleValue = value!!.toDouble()
-        val fitsScale = doubleValue != Double.POSITIVE_INFINITY
-                && doubleValue != Double.NEGATIVE_INFINITY
-                && value.compareTo(BigDecimal.valueOf(doubleValue)) == 0
-        return if (fitsScale) {
-            doubleValue
-        } else {
-            log.warn(errorMsg.format(key, "Double"))
-            value.toPlainString()
-        }
-    }
-
     companion object {
-        @JvmStatic private val UTC = ZoneId.of("UTC")
+        @JvmStatic
+        private val UTC = ZoneId.of("UTC")
     }
 
     override fun convert(data: Any?): MutableMap<String, Value?> {
-        val dataManipulatedForNeo4j = if(data is Map<*, *>)
+        val dataManipulatedForNeo4j = if (data is Map<*, *>) {
             data.mapValues {
-                when (val value = it.value) {
-                    is BigDecimal -> parseBigDecimal(it.key as String, value)
-                    is BigInteger -> {
-                        try {
-                            value.longValueExact()
-                        } catch (e: java.lang.ArithmeticException) {
-                            log.warn(errorMsg.format(it.key, "Long"))
-                            value.toString()
-                        }
+                val value = it.value
+                if (value is BigInteger) {
+                    try {
+                        value.longValueExact()
+                    } catch (e: java.lang.ArithmeticException) {
+                        log.warn(errorMsg.format(it.key, "Long"))
+                        value.toString()
                     }
-                    else -> value
+                } else {
+                    value
                 }
-            } else data
+            }
+        } else {
+            data
+        }
 
         return super.convert(dataManipulatedForNeo4j)
     }
+
 
     override fun setValue(result: MutableMap<String, Value?>?, fieldName: String?, value: Any?) {
         if (result != null && fieldName != null) {
@@ -63,7 +54,16 @@ class Neo4jValueConverter: MapValueConverter<Value>() {
     }
 
     override fun setDecimalField(result: MutableMap<String, Value?>?, fieldName: String?, value: BigDecimal?) {
-        setValue(result, fieldName, parseBigDecimal(fieldName, value))
+        val doubleValue = value!!.toDouble()
+        val fitsScale = doubleValue != Double.POSITIVE_INFINITY
+                && doubleValue != Double.NEGATIVE_INFINITY
+                && value.compareTo(BigDecimal.valueOf(doubleValue)) == 0
+        if (fitsScale) {
+            setValue(result, fieldName, doubleValue)
+        } else {
+            log.warn(errorMsg.format(fieldName, "Double"))
+            setValue(result, fieldName, value.toPlainString())
+        }
     }
 
     override fun setTimestampField(result: MutableMap<String, Value?>?, fieldName: String?, value: Date?) {

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
@@ -57,16 +57,41 @@ class Neo4jValueConverterTest {
     @Test
     fun `should convert tree with mixes types into map of neo4j values`() {
 
-        val result = Neo4jValueConverter().convert(Struct(TEST_SCHEMA)) as Map<*, *>
+        val utc = ZoneId.of("UTC")
+        val result = Neo4jValueConverter().convert(Struct(TEST_SCHEMA)) as Map<String, Value?>
 
-        assertTrue{ result["target"] is FloatValue }
-        assertTrue{ result["largeDecimal"] is StringValue }
-        assertTrue{ result["byteArray"] is BytesValue }
-        assertTrue{ result["int64"] is IntegerValue }
-        assertTrue{ result["int64Timestamp"] is LocalDateTimeValue }
-        assertTrue{ result["int32"] is IntegerValue }
-        assertTrue{ result["int32Date"] is DateValue }
-        assertTrue{ result["int32Time"] is LocalTimeValue }
+        val target = result["target"]
+        assertTrue{ target is FloatValue }
+        assertEquals(123.4, target?.asDouble())
+
+        val largeDecimal = result["largeDecimal"]
+        assertTrue{ largeDecimal is StringValue }
+        assertEquals(BigDecimal.valueOf(Double.MAX_VALUE).pow(2).toPlainString(), largeDecimal?.asString())
+
+        val byteArray = result["byteArray"]
+        assertTrue{ byteArray is BytesValue }
+        assertEquals("Foo".toByteArray().map { it }, byteArray?.asByteArray()?.map { it })
+
+        val int64 = result["int64"]
+        assertTrue{ int64 is IntegerValue }
+        assertEquals(Long.MAX_VALUE, int64?.asLong())
+
+        val int64Timestamp = result["int64Timestamp"]
+        assertTrue{ int64Timestamp is LocalDateTimeValue }
+        assertEquals(Date.from(Instant.ofEpochMilli(789)).toInstant().atZone(utc).toLocalDateTime(), int64Timestamp?.asLocalDateTime())
+
+        val int32 = result["int32"]
+        assertTrue{ int32 is IntegerValue }
+        assertEquals(123, int32?.asInt())
+
+        val int32Date = result["int32Date"]
+        assertTrue{ int32Date is DateValue }
+        assertEquals(Date.from(Instant.ofEpochMilli(456)).toInstant().atZone(utc).toLocalDate(), int32Date?.asLocalDate())
+
+        val int32Time = result["int32Time"]
+        assertTrue{ int32Time is LocalTimeValue }
+        assertEquals(Date.from(Instant.ofEpochMilli(123)).toInstant().atZone(utc).toLocalTime(), int32Time?.asLocalTime())
+
     }
 
     @Test
@@ -238,7 +263,7 @@ class Neo4jValueConverterTest {
                 .field("int64Timestamp",
                         ConnectSchema(Schema.Type.INT64,
                                 false,
-                                Date(),
+                                Date.from(Instant.ofEpochMilli(789)),
                                 Timestamp.LOGICAL_NAME,
                                 null, null))
                 .field("int32",
@@ -250,7 +275,7 @@ class Neo4jValueConverterTest {
                 .field("int32Date",
                         ConnectSchema(Schema.Type.INT32,
                                 false,
-                                Date(),
+                                Date.from(Instant.ofEpochMilli(456)),
                                 org.apache.kafka.connect.data.Date.LOGICAL_NAME,
                                 null, null))
                 .field("int32Time",

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
@@ -7,7 +7,12 @@ import org.junit.Test
 import org.neo4j.driver.Value
 import org.neo4j.driver.Values
 import streams.kafka.connect.sink.converters.Neo4jValueConverter
+import java.math.BigDecimal
+import java.math.BigInteger
+import java.time.ZoneId
+import java.util.*
 import kotlin.test.assertEquals
+import kotlin.test.assertTrue
 
 class Neo4jValueConverterTest {
 
@@ -47,6 +52,128 @@ class Neo4jValueConverterTest {
         // then
         val expected = getExpectedMap()
         assertEquals(expected, result)
+    }
+
+    @Test
+    fun `should convert BigInteger into String value if is less then Long MIN_VALUE`() {
+
+        val number = BigInteger.valueOf(Long.MIN_VALUE).pow(2)
+        val result = Neo4jValueConverter().convert(getItemElement(number))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertEquals(number.toString(), item?.asString())
+    }
+
+    @Test
+    fun `should convert BigInteger into String value if is greater then Long MAX_VALUE`() {
+
+        val number = BigInteger.valueOf(Long.MAX_VALUE).pow(2)
+        val result = Neo4jValueConverter().convert(getItemElement(number))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertEquals(number.toString(), item?.asString())
+    }
+
+    @Test
+    fun `should convert BigInteger into Long neo4j value if is equal to Long MAX_VALUE`() {
+
+        val number = Long.MAX_VALUE
+        val result = Neo4jValueConverter().convert(getItemElement(BigInteger.valueOf(number)))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.IntegerValue }
+        assertEquals(number, item?.asLong())
+    }
+
+    @Test
+    fun `should convert BigInteger into Long neo4j value if is less then Long MAX_VALUE and greater then Long MIN_VALUE`() {
+
+        val number: Long = 12345
+        val result = Neo4jValueConverter().convert(getItemElement(BigInteger.valueOf(number)))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.IntegerValue }
+        assertEquals(number, item?.asLong())
+    }
+
+    @Test
+    fun `should convert BigDecimal into String neo4j value if is a positive less then Double MIN_VALUE`() {
+
+        val number = BigDecimal.valueOf(Double.MIN_VALUE).pow(2)
+        val result = Neo4jValueConverter().convert(getItemElement(number))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertEquals(number.toPlainString(), item?.asString())
+    }
+
+    @Test
+    fun `should convert BigDecimal into String neo4j value if is a negative less then Double MAX_VALUE`() {
+
+        val number = - (BigDecimal.valueOf(Double.MAX_VALUE)).multiply(BigDecimal.valueOf(2))
+        val result = Neo4jValueConverter().convert(getItemElement(number))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertEquals(number.toPlainString(), item?.asString())
+    }
+
+    @Test
+    fun `should convert BigDecimal into String neo4j value if is greater then Double MAX_VALUE`() {
+
+        val number = BigDecimal.valueOf(Double.MAX_VALUE).pow(2)
+        val result = Neo4jValueConverter().convert(getItemElement(number))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertEquals(number.toPlainString(), item?.asString())
+    }
+
+    @Test
+    fun `should convert BigDecimal into Double neo4j value if is less then Double MAX_VALUE and greater then Double MIN_VALUE`() {
+
+        val number = 3456.78
+        val result =  Neo4jValueConverter().convert(getItemElement(BigDecimal.valueOf(number)))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.FloatValue }
+        assertEquals(number, item?.asDouble())
+    }
+
+    @Test
+    fun `should convert BigDecimal into Double neo4j value if is equal to Double MAX_VALUE`() {
+
+        val number = Double.MAX_VALUE
+        val result = Neo4jValueConverter().convert(getItemElement(BigDecimal.valueOf(number)))
+        val item = result["item"]
+
+        assertTrue{ item is org.neo4j.driver.internal.value.FloatValue }
+        assertEquals(number, item?.asDouble())
+    }
+
+
+    @Test
+    fun `should convert properly mixed items`() {
+
+        val double = Double.MAX_VALUE
+        val long = Long.MAX_VALUE
+        val bigDouble =  BigDecimal.valueOf(Double.MAX_VALUE).pow(2)
+        val string =  "FooBar"
+        val date = Date()
+        val result = Neo4jValueConverter().convert(mapOf(
+                "double" to double,
+                "long" to long,
+                "bigDouble" to bigDouble,
+                "string" to string,
+                "date" to date ))
+
+        assertEquals(double, result["double"]?.asDouble())
+        assertEquals(long, result["long"]?.asLong())
+        assertEquals(bigDouble.toPlainString(), result["bigDouble"]?.asString())
+        assertEquals(string, result["string"]?.asString())
+        assertEquals(date.toInstant().atZone(ZoneId.of("UTC")).toLocalDateTime(), result["date"]?.asLocalDateTime())
     }
 
     companion object {
@@ -113,6 +240,8 @@ class Neo4jValueConverterTest {
                     mapOf("value" to "Second Paragraph"))
             return mapOf("ul" to ulListMap, "p" to pListMap)
         }
+
+        fun getItemElement(number: Any): Map<String, Any> = mapOf("item" to number )
     }
 
 }

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
@@ -94,7 +94,7 @@ class Neo4jValueConverterTest {
     }
 
     @Test
-    fun `should convert BigDecimal into String neo4j value if is a positive less then Double MIN_VALUE`() {
+    fun `should convert BigDecimal into String neo4j value if is a positive less than Double MIN_VALUE`() {
 
         val number = BigDecimal.valueOf(Double.MIN_VALUE).pow(2)
         val result = Neo4jValueConverter().convert(getItemElement(number))
@@ -105,7 +105,7 @@ class Neo4jValueConverterTest {
     }
 
     @Test
-    fun `should convert BigDecimal into String neo4j value if is a negative less then Double MAX_VALUE`() {
+    fun `should convert BigDecimal into String neo4j value if is a negative less than Double MAX_VALUE`() {
 
         val number = - (BigDecimal.valueOf(Double.MAX_VALUE)).multiply(BigDecimal.valueOf(2))
         val result = Neo4jValueConverter().convert(getItemElement(number))
@@ -116,7 +116,7 @@ class Neo4jValueConverterTest {
     }
 
     @Test
-    fun `should convert BigDecimal into String neo4j value if is greater then Double MAX_VALUE`() {
+    fun `should convert BigDecimal into String neo4j value if is greater than Double MAX_VALUE`() {
 
         val number = BigDecimal.valueOf(Double.MAX_VALUE).pow(2)
         val result = Neo4jValueConverter().convert(getItemElement(number))
@@ -127,7 +127,7 @@ class Neo4jValueConverterTest {
     }
 
     @Test
-    fun `should convert BigDecimal into Double neo4j value if is less then Double MAX_VALUE and greater then Double MIN_VALUE`() {
+    fun `should convert BigDecimal into Double neo4j value if is less than Double MAX_VALUE and greater than Double MIN_VALUE`() {
 
         val number = 3456.78
         val result =  Neo4jValueConverter().convert(getItemElement(BigDecimal.valueOf(number)))

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
@@ -1,16 +1,16 @@
 package streams.kafka.connect.sink
 
-import org.apache.kafka.connect.data.Schema
-import org.apache.kafka.connect.data.SchemaBuilder
-import org.apache.kafka.connect.data.Struct
+import org.apache.kafka.connect.data.*
 import org.junit.Test
 import org.neo4j.driver.Value
 import org.neo4j.driver.Values
+import org.neo4j.driver.internal.value.*
 import streams.kafka.connect.sink.converters.Neo4jValueConverter
 import java.math.BigDecimal
 import java.math.BigInteger
+import java.time.Instant
 import java.time.ZoneId
-import java.util.*
+import java.util.Date
 import kotlin.test.assertEquals
 import kotlin.test.assertTrue
 
@@ -55,13 +55,28 @@ class Neo4jValueConverterTest {
     }
 
     @Test
+    fun `should convert tree with mixes types into map of neo4j values`() {
+
+        val result = Neo4jValueConverter().convert(Struct(TEST_SCHEMA)) as Map<*, *>
+
+        assertTrue{ result["target"] is FloatValue }
+        assertTrue{ result["largeDecimal"] is StringValue }
+        assertTrue{ result["byteArray"] is BytesValue }
+        assertTrue{ result["int64"] is IntegerValue }
+        assertTrue{ result["int64Timestamp"] is LocalDateTimeValue }
+        assertTrue{ result["int32"] is IntegerValue }
+        assertTrue{ result["int32Date"] is DateValue }
+        assertTrue{ result["int32Time"] is LocalTimeValue }
+    }
+
+    @Test
     fun `should convert BigInteger into String value if is less then Long MIN_VALUE`() {
 
         val number = BigInteger.valueOf(Long.MIN_VALUE).pow(2)
         val result = Neo4jValueConverter().convert(getItemElement(number))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertTrue{ item is StringValue }
         assertEquals(number.toString(), item?.asString())
     }
 
@@ -72,7 +87,7 @@ class Neo4jValueConverterTest {
         val result = Neo4jValueConverter().convert(getItemElement(number))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertTrue{ item is StringValue }
         assertEquals(number.toString(), item?.asString())
     }
 
@@ -83,7 +98,7 @@ class Neo4jValueConverterTest {
         val result = Neo4jValueConverter().convert(getItemElement(BigInteger.valueOf(number)))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.IntegerValue }
+        assertTrue{ item is IntegerValue }
         assertEquals(number, item?.asLong())
     }
 
@@ -94,7 +109,7 @@ class Neo4jValueConverterTest {
         val result = Neo4jValueConverter().convert(getItemElement(BigInteger.valueOf(number)))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.IntegerValue }
+        assertTrue{ item is IntegerValue }
         assertEquals(number, item?.asLong())
     }
 
@@ -105,7 +120,7 @@ class Neo4jValueConverterTest {
         val result = Neo4jValueConverter().convert(getItemElement(number))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertTrue{ item is StringValue }
         assertEquals(number.toPlainString(), item?.asString())
     }
 
@@ -116,7 +131,7 @@ class Neo4jValueConverterTest {
         val result = Neo4jValueConverter().convert(getItemElement(number))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertTrue{ item is StringValue }
         assertEquals(number.toPlainString(), item?.asString())
     }
 
@@ -127,7 +142,7 @@ class Neo4jValueConverterTest {
         val result = Neo4jValueConverter().convert(getItemElement(number))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.StringValue }
+        assertTrue{ item is StringValue }
         assertEquals(number.toPlainString(), item?.asString())
     }
 
@@ -138,7 +153,7 @@ class Neo4jValueConverterTest {
         val result =  Neo4jValueConverter().convert(getItemElement(BigDecimal.valueOf(number)))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.FloatValue }
+        assertTrue{ item is FloatValue }
         assertEquals(number, item?.asDouble())
     }
 
@@ -149,7 +164,7 @@ class Neo4jValueConverterTest {
         val result = Neo4jValueConverter().convert(getItemElement(BigDecimal.valueOf(number)))
         val item = result["item"]
 
-        assertTrue{ item is org.neo4j.driver.internal.value.FloatValue }
+        assertTrue{ item is FloatValue }
         assertEquals(number, item?.asDouble())
     }
 
@@ -167,7 +182,7 @@ class Neo4jValueConverterTest {
                 "long" to long,
                 "bigDouble" to bigDouble,
                 "string" to string,
-                "date" to date ))
+                "date" to date))
 
         assertEquals(double, result["double"]?.asDouble())
         assertEquals(long, result["long"]?.asLong())
@@ -193,6 +208,57 @@ class Neo4jValueConverterTest {
         private val BODY_SCHEMA = SchemaBuilder.struct().name("org.neo4j.example.html.BODY")
                 .field("ul", SchemaBuilder.array(UL_SCHEMA).optional())
                 .field("p", SchemaBuilder.array(P_SCHEMA).optional())
+                .build()
+
+        private val TEST_SCHEMA = SchemaBuilder.struct().name("test.schema")
+                .field("target",
+                        ConnectSchema(Schema.Type.BYTES,
+                                false,
+                                BigDecimal.valueOf(123.4),
+                                Decimal.LOGICAL_NAME,
+                                null, null))
+                .field("largeDecimal",
+                        ConnectSchema(Schema.Type.BYTES,
+                                false,
+                                BigDecimal.valueOf(Double.MAX_VALUE).pow(2),
+                                Decimal.LOGICAL_NAME,
+                                null, null))
+                .field("byteArray",
+                        ConnectSchema(Schema.Type.BYTES,
+                                false,
+                                "Foo".toByteArray(),
+                                "name.byteArray",
+                                null, null))
+                .field("int64",
+                        ConnectSchema(Schema.Type.INT64,
+                                false,
+                                Long.MAX_VALUE,
+                                "name.int64",
+                                null, null))
+                .field("int64Timestamp",
+                        ConnectSchema(Schema.Type.INT64,
+                                false,
+                                Date(),
+                                Timestamp.LOGICAL_NAME,
+                                null, null))
+                .field("int32",
+                        ConnectSchema(Schema.Type.INT32,
+                                false,
+                                123,
+                                "name.int32",
+                                null, null))
+                .field("int32Date",
+                        ConnectSchema(Schema.Type.INT32,
+                                false,
+                                Date(),
+                                org.apache.kafka.connect.data.Date.LOGICAL_NAME,
+                                null, null))
+                .field("int32Time",
+                        ConnectSchema(Schema.Type.INT32,
+                                false,
+                                Date.from(Instant.ofEpochMilli(123)),
+                                Time.LOGICAL_NAME,
+                                null, null))
                 .build()
 
         fun getTreeStruct(): Struct? {
@@ -241,7 +307,7 @@ class Neo4jValueConverterTest {
             return mapOf("ul" to ulListMap, "p" to pListMap)
         }
 
-        fun getItemElement(number: Any): Map<String, Any> = mapOf("item" to number )
+        fun getItemElement(number: Any): Map<String, Any> = mapOf("item" to number)
     }
 
 }

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
@@ -193,7 +193,6 @@ class Neo4jValueConverterTest {
         assertEquals(number, item?.asDouble())
     }
 
-
     @Test
     fun `should convert properly mixed items`() {
 

--- a/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
+++ b/kafka-connect-neo4j/src/test/kotlin/streams/kafka/connect/sink/Neo4jValueConverterTest.kt
@@ -7,7 +7,6 @@ import org.neo4j.driver.Values
 import org.neo4j.driver.internal.value.*
 import streams.kafka.connect.sink.converters.Neo4jValueConverter
 import java.math.BigDecimal
-import java.math.BigInteger
 import java.time.Instant
 import java.time.ZoneId
 import java.util.Date
@@ -92,50 +91,6 @@ class Neo4jValueConverterTest {
         assertTrue{ int32Time is LocalTimeValue }
         assertEquals(Date.from(Instant.ofEpochMilli(123)).toInstant().atZone(utc).toLocalTime(), int32Time?.asLocalTime())
 
-    }
-
-    @Test
-    fun `should convert BigInteger into String value if is less then Long MIN_VALUE`() {
-
-        val number = BigInteger.valueOf(Long.MIN_VALUE).pow(2)
-        val result = Neo4jValueConverter().convert(getItemElement(number))
-        val item = result["item"]
-
-        assertTrue{ item is StringValue }
-        assertEquals(number.toString(), item?.asString())
-    }
-
-    @Test
-    fun `should convert BigInteger into String value if is greater then Long MAX_VALUE`() {
-
-        val number = BigInteger.valueOf(Long.MAX_VALUE).pow(2)
-        val result = Neo4jValueConverter().convert(getItemElement(number))
-        val item = result["item"]
-
-        assertTrue{ item is StringValue }
-        assertEquals(number.toString(), item?.asString())
-    }
-
-    @Test
-    fun `should convert BigInteger into Long neo4j value if is equal to Long MAX_VALUE`() {
-
-        val number = Long.MAX_VALUE
-        val result = Neo4jValueConverter().convert(getItemElement(BigInteger.valueOf(number)))
-        val item = result["item"]
-
-        assertTrue{ item is IntegerValue }
-        assertEquals(number, item?.asLong())
-    }
-
-    @Test
-    fun `should convert BigInteger into Long neo4j value if is less then Long MAX_VALUE and greater then Long MIN_VALUE`() {
-
-        val number: Long = 12345
-        val result = Neo4jValueConverter().convert(getItemElement(BigInteger.valueOf(number)))
-        val item = result["item"]
-
-        assertTrue{ item is IntegerValue }
-        assertEquals(number, item?.asLong())
     }
 
     @Test

